File: flutter/vehicle_debug_ui.dart
import 'package:flutter/material.dart';
import 'pid_router.dart';
import 'package:intl/intl.dart';

/// Araç Debug Paneli - Tam panel
class VehicleDebugPanel extends StatefulWidget {
  final PIDRouter pidRouter;
  final VoidCallback onClose;

  const VehicleDebugPanel({
    Key? key,
    required this.pidRouter,
    required this.onClose,
  }) : super(key: key);

  @override
  State<VehicleDebugPanel> createState() => _VehicleDebugPanelState();
}

class _VehicleDebugPanelState extends State<VehicleDebugPanel> {
  VehicleDetectionResult? _detectionResult;
  bool _isExpanded = true;
  bool _isLoading = true;

  @override
  void initState() {
    super.initState();
    _detectVehicle();
  }

  Future<void> _detectVehicle() async {
    final result = await widget.pidRouter.detectVehicle();
    setState(() {
      _detectionResult = result;
      _isLoading = false;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Card(
      elevation: 8,
      shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(12)),
      child: Container(
        constraints: const BoxConstraints(maxWidth: 400),
        padding: const EdgeInsets.all(12),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            // Header
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                Row(
                  children: [
                    const Icon(Icons.build, color: Colors.blue, size: 20),
                    const SizedBox(width: 8),
                    Text(
                      'Vehicle Detection Debug',
                      style: Theme.of(context).textTheme.titleMedium?.copyWith(
                            fontWeight: FontWeight.bold,
                          ),
                    ),
                  ],
                ),
                Row(
                  children: [
                    IconButton(
                      icon: Icon(_isExpanded
                          ? Icons.keyboard_arrow_up
                          : Icons.keyboard_arrow_down),
                      onPressed: () {
                        setState(() {
                          _isExpanded = !_isExpanded;
                        });
                      },
                    ),
                    IconButton(
                      icon: const Icon(Icons.close),
                      onPressed: widget.onClose,
                    ),
                  ],
                ),
              ],
            ),

            if (_isExpanded) ...[
              const SizedBox(height: 12),

              // Content
              if (_isLoading)
                const Padding(
                  padding: EdgeInsets.all(24),
                  child: CircularProgressIndicator(),
                )
              else if (_detectionResult != null) ...[
                _buildBrandDisplay(_detectionResult!),
                const SizedBox(height: 12),
                _buildDetectionMethods(_detectionResult!),
                const SizedBox(height: 12),
                _buildStats(_detectionResult!),
              ],
            ],
          ],
        ),
      ),
    );
  }

  Widget _buildBrandDisplay(VehicleDetectionResult result) {
    final color = _getColorForConfidence(result.confidence);

    return Container(
      width: double.infinity,
      decoration: BoxDecoration(
        color: color,
        borderRadius: BorderRadius.circular(8),
      ),
      padding: const EdgeInsets.all(16),
      child: Column(
        children: [
          const Icon(Icons.directions_car, size: 48, color: Colors.white),
          const SizedBox(height: 8),
          Text(
            result.brand.displayName,
            style: const TextStyle(
              fontSize: 24,
              fontWeight: FontWeight.bold,
              color: Colors.white,
            ),
          ),
          if (result.model != null)
            Text(
              'Model: ${result.model}',
              style: const TextStyle(fontSize: 14, color: Colors.white70),
            ),
          const SizedBox(height: 4),
          Text(
            '${result.confidence}% Confidence',
            style: const TextStyle(fontSize: 12, color: Colors.white60),
          ),
        ],
      ),
    );
  }

  Widget _buildDetectionMethods(VehicleDetectionResult result) {
    return Card(
      color: Colors.grey[100],
      child: Padding(
        padding: const EdgeInsets.all(12),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text(
              'Detection Methods',
              style: Theme.of(context).textTheme.titleSmall?.copyWith(
                    fontWeight: FontWeight.bold,
                  ),
            ),
            const SizedBox(height: 8),
            if (result.detectionMethods.isEmpty)
              const Text(
                'No detection methods available',
                style: TextStyle(color: Colors.grey),
              )
            else
              ...result.detectionMethods.map((method) {
                return Padding(
                  padding: const EdgeInsets.only(bottom: 4),
                  child: _buildDetectionMethodItem(method),
                );
              }).toList(),
          ],
        ),
      ),
    );
  }

  Widget _buildDetectionMethodItem(DetectionMethod method) {
    IconData icon;
    switch (method.source) {
      case 'VIN':
        icon = Icons.info;
        break;
      case 'ECU':
        icon = Icons.settings;
        break;
      case 'PID Test':
        icon = Icons.play_arrow;
        break;
      default:
        icon = Icons.help;
    }

    return Container(
      decoration: BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.circular(4),
      ),
      padding: const EdgeInsets.all(8),
      child: Row(
        children: [
          Icon(icon, size: 16, color: Colors.blue),
          const SizedBox(width: 8),
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(
                  method.source,
                  style: const TextStyle(
                    fontSize: 12,
                    fontWeight: FontWeight.w500,
                  ),
                ),
                if (method.data != null)
                  Text(
                    method.data!,
                    style: const TextStyle(fontSize: 10, color: Colors.grey),
                  ),
              ],
            ),
          ),
          _buildConfidenceBadge(method.confidence),
        ],
      ),
    );
  }

  Widget _buildConfidenceBadge(int confidence) {
    final color = _getColorForConfidence(confidence);
    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 4),
      decoration: BoxDecoration(
        color: color,
        borderRadius: BorderRadius.circular(12),
      ),
      child: Text(
        '$confidence%',
        style: const TextStyle(
          fontSize: 10,
          fontWeight: FontWeight.bold,
          color: Colors.white,
        ),
      ),
    );
  }

  Widget _buildStats(VehicleDetectionResult result) {
    final dateFormat = DateFormat('HH:mm:ss');
    final timeStr =
        dateFormat.format(DateTime.fromMillisecondsSinceEpoch(result.timestamp));

    return Container(
      decoration: BoxDecoration(
        color: Colors.grey[100],
        borderRadius: BorderRadius.circular(4),
      ),
      padding: const EdgeInsets.all(8),
      child: Column(
        children: [
          _buildStatRow('Detection Time', timeStr),
          _buildStatRow('Methods Used', '${result.detectionMethods.length}'),
          _buildStatRow('Brand Code', result.brand.name.toUpperCase()),
        ],
      ),
    );
  }

  Widget _buildStatRow(String label, String value) {
    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 2),
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: [
          Text(
            label,
            style: const TextStyle(fontSize: 11, color: Colors.grey),
          ),
          Text(
            value,
            style: const TextStyle(
              fontSize: 11,
              fontWeight: FontWeight.w500,
            ),
          ),
        ],
      ),
    );
  }

  Color _getColorForConfidence(int confidence) {
    if (confidence >= 90) return Colors.green;
    if (confidence >= 70) return Colors.blue;
    if (confidence >= 50) return Colors.amber;
    return Colors.red;
  }
}

/// Araç Debug Rozeti - Kompakt
class VehicleDebugBadge extends StatefulWidget {
  final PIDRouter pidRouter;
  final VoidCallback onClick;

  const VehicleDebugBadge({
    Key? key,
    required this.pidRouter,
    required this.onClick,
  }) : super(key: key);

  @override
  State<VehicleDebugBadge> createState() => _VehicleDebugBadgeState();
}

class _VehicleDebugBadgeState extends State<VehicleDebugBadge> {
  VehicleBrand? _brand;
  int _confidence = 0;

  @override
  void initState() {
    super.initState();
    _loadBrand();
  }

  Future<void> _loadBrand() async {
    final result = await widget.pidRouter.detectVehicle();
    setState(() {
      _brand = result.brand;
      _confidence = result.confidence;
    });
  }

  @override
  Widget build(BuildContext context) {
    final color = _brand != null ? _getColorForConfidence(_confidence) : Colors.grey;

    return InkWell(
      onTap: widget.onClick,
      child: Container(
        padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 6),
        decoration: BoxDecoration(
          color: color,
          borderRadius: BorderRadius.circular(16),
        ),
        child: Row(
          mainAxisSize: MainAxisSize.min,
          children: [
            const Icon(Icons.directions_car, size: 16, color: Colors.white),
            const SizedBox(width: 6),
            Text(
              _brand?.displayName ?? 'Detecting...',
              style: const TextStyle(
                fontSize: 12,
                fontWeight: FontWeight.w500,
                color: Colors.white,
              ),
            ),
          ],
        ),
      ),
    );
  }

  Color _getColorForConfidence(int confidence) {
    if (confidence >= 90) return Colors.green;
    if (confidence >= 70) return Colors.blue;
    if (confidence >= 50) return Colors.amber;
    return Colors.red;
  }
}

File: VehicleDebugUI.kt
package com.obd.diagnostics.ui

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.obd.diagnostics.router.PIDRouter
import java.text.SimpleDateFormat
import java.util.*

/**
 * Debug UI component to display vehicle brand detection
 * Can be toggled on/off and positioned as floating overlay
 */
@Composable
fun VehicleDebugPanel(
    pidRouter: PIDRouter,
    modifier: Modifier = Modifier,
    onClose: () -> Unit = {}
) {
    val detectionResult = remember { mutableStateOf<PIDRouter.VehicleDetectionResult?>(null) }
    val isExpanded = remember { mutableStateOf(true) }
    
    LaunchedEffect(Unit) {
        // Detect vehicle on component mount
        detectionResult.value = pidRouter.detectVehicle()
    }
    
    Card(
        modifier = modifier
            .padding(8.dp)
            .widthIn(max = 400.dp),
        elevation = CardDefaults.cardElevation(defaultElevation = 8.dp),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.surface.copy(alpha = 0.95f)
        ),
        shape = RoundedCornerShape(12.dp)
    ) {
        Column(
            modifier = Modifier.padding(12.dp)
        ) {
            // Header
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Icon(
                        imageVector = Icons.Default.Build,
                        contentDescription = "Debug",
                        tint = MaterialTheme.colorScheme.primary,
                        modifier = Modifier.size(20.dp)
                    )
                    Spacer(modifier = Modifier.width(8.dp))
                    Text(
                        text = "Vehicle Detection Debug",
                        style = MaterialTheme.typography.titleMedium,
                        fontWeight = FontWeight.Bold
                    )
                }
                Row {
                    IconButton(
                        onClick = { isExpanded.value = !isExpanded.value },
                        modifier = Modifier.size(32.dp)
                    ) {
                        Icon(
                            imageVector = if (isExpanded.value) Icons.Default.KeyboardArrowUp 
                                         else Icons.Default.KeyboardArrowDown,
                            contentDescription = "Expand/Collapse"
                        )
                    }
                    IconButton(
                        onClick = onClose,
                        modifier = Modifier.size(32.dp)
                    ) {
                        Icon(
                            imageVector = Icons.Default.Close,
                            contentDescription = "Close"
                        )
                    }
                }
            }
            
            AnimatedVisibility(
                visible = isExpanded.value,
                enter = fadeIn(),
                exit = fadeOut()
            ) {
                detectionResult.value?.let { result ->
                    Column(modifier = Modifier.padding(top = 12.dp)) {
                        // Main brand display
                        VehicleBrandDisplay(result)
                        
                        Spacer(modifier = Modifier.height(12.dp))
                        
                        // Detection methods
                        DetectionMethodsList(result)
                        
                        Spacer(modifier = Modifier.height(12.dp))
                        
                        // Stats
                        DetectionStats(result)
                    }
                } ?: run {
                    // Loading state
                    Box(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(24.dp),
                        contentAlignment = Alignment.Center
                    ) {
                        CircularProgressIndicator()
                    }
                }
            }
        }
    }
}

@Composable
private fun VehicleBrandDisplay(result: PIDRouter.VehicleDetectionResult) {
    Surface(
        modifier = Modifier.fillMaxWidth(),
        color = getColorForConfidence(result.confidence),
        shape = RoundedCornerShape(8.dp)
    ) {
        Column(
            modifier = Modifier.padding(16.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Icon(
                imageVector = Icons.Default.DirectionsCar,
                contentDescription = "Vehicle",
                modifier = Modifier.size(48.dp),
                tint = Color.White
            )
            Spacer(modifier = Modifier.height(8.dp))
            Text(
                text = result.brand.displayName,
                fontSize = 24.sp,
                fontWeight = FontWeight.Bold,
                color = Color.White
            )
            result.model?.let { model ->
                Text(
                    text = "Model: $model",
                    fontSize = 14.sp,
                    color = Color.White.copy(alpha = 0.9f)
                )
            }
            Spacer(modifier = Modifier.height(4.dp))
            Text(
                text = "${result.confidence}% Confidence",
                fontSize = 12.sp,
                color = Color.White.copy(alpha = 0.8f)
            )
        }
    }
}

@Composable
private fun DetectionMethodsList(result: PIDRouter.VehicleDetectionResult) {
    Card(
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.surfaceVariant
        )
    ) {
        Column(modifier = Modifier.padding(12.dp)) {
            Text(
                text = "Detection Methods",
                style = MaterialTheme.typography.titleSmall,
                fontWeight = FontWeight.Bold
            )
            Spacer(modifier = Modifier.height(8.dp))
            
            if (result.detectionMethods.isEmpty()) {
                Text(
                    text = "No detection methods available",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.6f)
                )
            } else {
                result.detectionMethods.forEach { method ->
                    DetectionMethodItem(method)
                    Spacer(modifier = Modifier.height(4.dp))
                }
            }
        }
    }
}

@Composable
private fun DetectionMethodItem(method: PIDRouter.DetectionMethod) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .background(
                color = MaterialTheme.colorScheme.surface,
                shape = RoundedCornerShape(4.dp)
            )
            .padding(8.dp),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Icon(
                imageVector = when (method) {
                    is PIDRouter.DetectionMethod.VIN -> Icons.Default.Info
                    is PIDRouter.DetectionMethod.ECU -> Icons.Default.Settings
                    is PIDRouter.DetectionMethod.PIDTest -> Icons.Default.PlayArrow
                },
                contentDescription = method.source,
                modifier = Modifier.size(16.dp),
                tint = MaterialTheme.colorScheme.primary
            )
            Spacer(modifier = Modifier.width(8.dp))
            Column {
                Text(
                    text = method.source,
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Medium
                )
                when (method) {
                    is PIDRouter.DetectionMethod.VIN -> {
                        Text(
                            text = "VIN: ${method.vin}",
                            fontSize = 10.sp,
                            color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.6f)
                        )
                    }
                    is PIDRouter.DetectionMethod.ECU -> {
                        Text(
                            text = "ECU: ${method.ecuName}",
                            fontSize = 10.sp,
                            color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.6f)
                        )
                    }
                    is PIDRouter.DetectionMethod.PIDTest -> {
                        Text(
                            text = "Result: ${method.testResult}",
                            fontSize = 10.sp,
                            color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.6f)
                        )
                    }
                }
            }
        }
        ConfidenceBadge(method.confidence)
    }
}

@Composable
private fun ConfidenceBadge(confidence: Int) {
    Surface(
        color = getColorForConfidence(confidence),
        shape = RoundedCornerShape(12.dp)
    ) {
        Text(
            text = "$confidence%",
            fontSize = 10.sp,
            fontWeight = FontWeight.Bold,
            color = Color.White,
            modifier = Modifier.padding(horizontal = 8.dp, vertical = 4.dp)
        )
    }
}

@Composable
private fun DetectionStats(result: PIDRouter.VehicleDetectionResult) {
    val dateFormat = remember { SimpleDateFormat("HH:mm:ss", Locale.getDefault()) }
    val timeStr = remember(result.timestamp) {
        dateFormat.format(Date(result.timestamp))
    }
    
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(
                color = MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.5f),
                shape = RoundedCornerShape(4.dp)
            )
            .padding(8.dp)
    ) {
        StatRow("Detection Time", timeStr)
        StatRow("Methods Used", "${result.detectionMethods.size}")
        StatRow("Brand Code", result.brand.name)
    }
}

@Composable
private fun StatRow(label: String, value: String) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween
    ) {
        Text(
            text = label,
            fontSize = 11.sp,
            color = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.7f)
        )
        Text(
            text = value,
            fontSize = 11.sp,
            fontWeight = FontWeight.Medium,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )
    }
}

private fun getColorForConfidence(confidence: Int): Color {
    return when {
        confidence >= 90 -> Color(0xFF4CAF50) // Green
        confidence >= 70 -> Color(0xFF2196F3) // Blue
        confidence >= 50 -> Color(0xFFFFC107) // Amber
        else -> Color(0xFFF44336) // Red
    }
}

/**
 * Minimal compact version for status bar
 */
@Composable
fun VehicleDebugBadge(
    pidRouter: PIDRouter,
    onClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    val brand = pidRouter.detectedBrand
    val confidence = remember { mutableStateOf(0) }
    
    LaunchedEffect(brand) {
        if (brand != null) {
            val result = pidRouter.detectVehicle()
            confidence.value = result.confidence
        }
    }
    
    Surface(
        modifier = modifier
            .clickable(onClick = onClick)
            .padding(4.dp),
        color = brand?.let { getColorForConfidence(confidence.value) } 
            ?: MaterialTheme.colorScheme.surfaceVariant,
        shape = RoundedCornerShape(16.dp)
    ) {
        Row(
            modifier = Modifier.padding(horizontal = 12.dp, vertical = 6.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Icon(
                imageVector = Icons.Default.DirectionsCar,
                contentDescription = "Vehicle",
                modifier = Modifier.size(16.dp),
                tint = Color.White
            )
            Spacer(modifier = Modifier.width(6.dp))
            Text(
                text = brand?.displayName ?: "Detecting...",
                fontSize = 12.sp,
                fontWeight = FontWeight.Medium,
                color = Color.White
            )
        }
    }
}
